### ISSUE TITLE
initShim breaks Spring Data JDBC

### ISSUE THREAD

**Issue by heldev** [immutables/immutables#1260]
Spring folks ignore the `transient` keyword [by design](https://jira.spring.io/browse/DATACMNS-1477?focusedCommentId=182335&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-182335). As a result Spring tries to persist the field and fails. 

I tried (without success) adding `@Transient` to it like this 

	@InjectTransient
	@Nullable
	protected abstract Object initShim();

	@InjectAnnotation(type = Transient.class, target = FIELD)
	@Target(METHOD)
	@interface InjectTransient {}

Would be nice to have a solution/workaround

**Comment by elucash** [comment]
Need some time to comprehend all implications, but can you please try the workaround (if it works, not much of a guarantee) when use old style default generation in form of  `@Style(unsafeDefaultAndDerived = true)`

**Comment by elucash** [comment]
Another consideration: have you tried `SYNTHETIC_FIELDS`?
1. Use `@InjectAnnotation(type = Transient.class, target = SYNTHETIC_FIELDS)`
2. Put `@InjectTransient` on the type, not on the method
3. Don't declare `abstract Object initShim()`, remove it

```java
@InjectAnnotation(type = Transient.class, target = SYNTHETIC_FIELDS)
@Target({TYPE, PACKAGE, ANNOTATION_TYPE})
@interface InjectTransient {}

@Value.Immutable
@InjectTransient  //<-- can put it as meta annotation, or on the parent package package-info.java
interface MyObj {}
```

**Comment by heldev** [comment]
Today I tried to reproduce the problem I had back then, and I couldn't. I'm able to save Immutables generated objects in a simple spring data JDBC hello wold project. My assumption is it was some kind of a weird side effect of other Spring bugs. The project used an uncommon config with a MySQL autoincrement outside of a primary key. Spring had difficulties with such a setup.
Even If I stumble upon the problem again the `SYNTHETIC_FIELDS` suggestion seems like a great solution. Thanks for a quick reply

**Comment by heldev** [comment]
I found how to reproduce it, just added `@VavrEncodingEnabled` and an `Option`-typed field, neither `unsafeDefaultAndDerived` nor `SYNTHETIC_FIELDS` helped

**Comment by aurambaj** [comment]
I have a similar issue while working with Spring jdbc. Got a working solution relying on `unsafeDefaultAndDerived = true`. @elucash Is that going to be an issue moving forward since it is the older way of doing things? 

(no luck with `SYNTHETIC_FIELDS` but I may have missed something)

```
@Immutable
@Value.Style(
    unsafeDefaultAndDerived = true,
    builtinContainerAttributes = false,
    typeImmutable = "*",
    of="new"
)
@InjectPersistenceConstructor
public abstract class AbstractOrganization implements Persistable<UUID> {

  @Value.Parameter
  @Id
  @Default
  public UUID getId() {
    return UUID.randomUUID();
  }

  @Value.Parameter
  @Default
  public Instant getCreatedAt() {
    return Instant.now();
  }

  @Value.Parameter
  public abstract String getName();

  /** for spring jdbc to control new instances if using manually set ids */
  @InjectTransient
  @Default
  @Override
  public boolean isNew() {
    return true;
  }
}```

**Comment by elucash** [comment]
just to clarify, `SYNTHETIC_FIELDS` should go with annotation covering whole type, just like `@InjectPersistenceConstructor` in the example above

**Comment by elucash** [comment]
`unsafeDefaultAndDerived` doesn't generate `initShim`, without it any cycles (calls within initializer methods `@Default` `@Derived`) are not detected so things can be `null` or constructor will blow with infinite recursion. Init Shim is transient, but as spring data prefers using `@Transient` annotation you can inject it on all synthetic fields (all which are not holding actual data, but generated for internal machinery, such as bit fields in builder or lazy flags in instances or the initShim. As I'm checking the codegen template to confirm if it's a bug, can you try to introduce `@Value.Lazy` attribute, then inject `@Transient` annotations on `SYNTHETIC_FIELDS` and watch lazy bits will have injected annotation applied? If there's no such annotation injected for the `initShim`, then there's missing application of injected annotation for it, we'll fix it.

**Comment by aurambaj** [comment]
Yes, I tried using `SYNTHETIC_FIELDS` but that doesn't inject `@Transient`. That

**Comment by elucash** [comment]
just curious, if `builtinContainerAttributes=true` will it generate proper array equals (or alternatively - reference equality)

**Comment by aurambaj** [comment]
yes, with `builtinContainerAttributes=true` it does use `Arrays.equals`

**Comment by elucash** [comment]
by disabling builtin container handling, Immutables doesnt not consider array as an array but as regular reference type (as anything extending Object) and calls `Object.equals`. You've probably disabled containers to fight something in a way containers work by default, so some other solution can be needed, dunno if this can be solved by `jdkOnly` or `beanFriendlyModifiables`. In general, if not for the peculiarities of serialization mechanisms, I would wrap any raw array into either collection or specialized type (`ByteString` / `MyBlob` / `Data`) just for the purpose of controlling `toString`, equivalence and [de-]serialization  (how hard it can be to override some serializers/adapters? well for some technologies this definitely a PITA)
